### PR TITLE
[plugin-applitools] Support applitools accessibility testing

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/converter/StringToAccessibilitySettingsConverter.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    public StringToAccessibilitySettingsConverter(FluentTrimmedEnumConverter fluentTrimmedEnumConverter)
+    {
+        this.fluentTrimmedEnumConverter = fluentTrimmedEnumConverter;
+    }
+
+    @Override
+    public AccessibilitySettings convertValue(String value, Type type)
+    {
+        Matcher matcher = ACCESSIBILITY_PATTERN.matcher(value);
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/converter/StringToAccessibilitySettingsConverter.java:44]: why not to split by first/last `-` ?

File: vividus-plugin-applitools/src/main/resources/ufg-applitools-visual-comparison.ftl [hunk 2]
```diff
@@ -43,6 +45,46 @@
                 </#list>
             </tbody>
         </table>
+        <#if testResults[0].getAccessibilityCheckResult()??>
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/resources/ufg-applitools-visual-comparison.ftl:48]: why is it a separate section?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/converter/StringToAccessibilitySettingsConverter.java (hunk 1 above)
Comment (uarlouski) [reply vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/converter/StringToAccessibilitySettingsConverter.java:44]: just used to use matcher for such cases, but agree - look simpler and faster

File: docs/modules/plugins/partials/applitools-configuration.adoc [hunk 3]
```diff
@@ -14,3 +14,4 @@
 ** `readApiKey` - The key to execute tests.
 ** `executeApiKey` - The key to access tests and test results.
 ** `serverUri` - The URI of the rest server.
+** `accessibilityStandard` - The accessibility standard to check the page against. The required value format is `<standard> - <level>`, where standard is either `WCAG 2.0` or `WCAG 2.1` and level is either `AA` or `AAA`. This check uses https://applitools.com/contrast-advisor/[Applitools Contrast Advisor] that helps in identifying potential contrast violations in both the text and images of your web and mobile applications, helping to ensure your technology is more accessible.
```
[the remark this answers is not in the document]
Comment (uarlouski) [reply docs/modules/plugins/partials/applitools-configuration.adoc:17]: nope, only AA and AAA
image

File: vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/AbstractApplitoolsSteps.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        super(uiContext, attachmentPublisher, softAssert);
+    }
+
+    protected void verifyAccessibility(ApplitoolsTestResults results)
+    {
+        AccessibilityCheckResult accessibilityStatus = results.getAccessibilityCheckResult();
+        if (accessibilityStatus != null)
+        {
+            getSoftAssert().assertTrue(
+                    String.format("The %s %s accessibility check for test: %s", accessibilityStatus.getVersion(),
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/AbstractApplitoolsSteps.java:40]:
```suggestion
                    String.format("%s - %s accessibility check for test: %s", accessibilityStatus.getVersion(),
```
it's needed to be consistent everywhere

File: vividus-plugin-applitools/src/main/resources/ufg-applitools-visual-comparison.ftl [hunk 5]
```diff
@@ -7,41 +7,63 @@
...
+                                    <td class="inner-cell">
+                                        Visual check is ${testResult.getStatus()}, see<a href="${testResult.getUrl()}" class="btn btn-link btn-sm active" role="button">Applitools report</a>for details.
+                                    </td>
+                                </tr>
+                                <#if testResult.getAccessibilityCheckResult()??>
+                                    <#assign accessibilityResult = testResult.getAccessibilityCheckResult()>
+                                    <#assign accessibilityStatus = accessibilityResult.isPassed()?then('success', 'failed')>
+                                    <tr class="${accessibilityStatus}">
+                                        <td class="inner-cell">
+                                            Accessibility WCAG 2.1 - AA check is ${accessibilityResult.getStatus()}, see<a href="${accessibilityResult.getUrl()}" class="btn btn-link btn-sm active" role="button">Applitools report</a>for details.
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/resources/ufg-applitools-visual-comparison.ftl:58]: > WCAG 2.1 - AA 
😮

File: vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/model/ApplitoolsTestResults.java [hunk 6]
```diff
@@ -30,13 +37,12 @@
     private final String os;
     private final String browser;
     private final RectangleSize viewport;
-    private final String device;
+    private String device = "";
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/model/ApplitoolsTestResults.java:40]:
```suggestion
    private Optional<String> device = Optional.empty();
```
?

**Comment by ikalinin1** [comment]
@uarlouski  In general LGTM! I think it makes sense to add examples of accessibility usage in the docs

**Review by ikalinin1 — APPROVED** [review]

File: vividus-plugin-applitools/src/main/resources/applitools-visual-comparison.ftl [hunk 7]
```diff
@@ -4,7 +4,37 @@
...
+
+.dot {
+  height: 13px;
+  width: 13px;
+  border-radius: 50%;
+  display: inline-block;
+}
+
+.red {
+  background-color: #dc3545s;
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/resources/applitools-visual-comparison.ftl:16]: something strange

File: vividus-tests/src/main/resources/story/system/UltrafastGridSteps.story [hunk 8]
```diff
@@ -51,11 +51,11 @@ When I run visual test with Applitools UFG using:
...
 
 
-Scenario: Validation of step: 'When I run visual test with Applitools UFG using:$applitoolsConfigurations and matrix:$matrix' for context element with element cut
+Scenario: Validation of step: 'When I run visual test with Applitools UFG using:$applitoolsConfigurations and matrix:$matrix' for context element with element cut and Accessibility check
 When I change context to element located by `xpath(.//body)`
 When I run visual test with Applitools UFG using:
-|baselineName            |batchName  |action  |elementsToIgnore|
-|ufg-context-element-cut |<batchName>|<action>|tagName(img)    |
+|baselineName            |batchName  |action  |elementsToIgnore|accessibilityStandard|
+|ufg-context-element-cut |<batchName>|<action>|tagName(img)    |WCAG 2.1 - AA        |
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/system/UltrafastGridSteps.story:58]: how can we be sure a11y is verified actually?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-tests/src/main/resources/story/system/UltrafastGridSteps.story (hunk 8 above)
Comment (uarlouski) [reply vividus-tests/src/main/resources/story/system/UltrafastGridSteps.story:58]: added known issue for failed check

File: docs/modules/plugins/pages/plugin-applitools.adoc [hunk 9]
```diff
@@ -259,3 +259,16 @@ When I run visual test with Applitools UFG using:
 |ios                    |       |            |iPhone X  |portrait         |latest |
 |chrome_mobile_emulation|       |            |Galaxy S10|landscape        |       |
 ----
+
+.Perform Perform Ultrafast Grid check combined with Accessibility check
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-applitools.adoc:263]:
```suggestion
.Perform Ultrafast Grid check combined with Accessibility check
```

File: vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/model/AccessibilityCheckResult.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+import org.apache.hc.core5.net.URIBuilder;
+
+public class AccessibilityCheckResult
+{
+    private static final String GUIDELINE_FORMAT = "%s - %s";
+
+    private final String url;
+    private final AccessibilityStatus status;
+    private final String guideline;
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/model/AccessibilityCheckResult.java:32]:
```suggestion
    private final String standard;
```
?

File: vividus-plugin-applitools/src/main/resources/applitools-visual-comparison.ftl [hunk 11]
```diff
@@ -4,7 +4,28 @@
...
+  width: 13px;
+  border-radius: 50%;
+  display: inline-block;
+}
 </#macro>
+
+<#macro custom_section>
+ <#if result.applitoolsTestResults.accessibilityCheckResult??>
+     <#assign checkResult = result.applitoolsTestResults.accessibilityCheckResult>
+     <#assign statusColor = checkResult.isPassed()?then('green', 'red')>
```
Comment (valfirst) [inline vividus-plugin-applitools/src/main/resources/applitools-visual-comparison.ftl:19]: I believe it was agreement to use common styles (including colors) for all attachments

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/model/AccessibilityCheckResult.java (hunk 10 above)
Comment (uarlouski) [reply vividus-plugin-applitools/src/main/java/org/vividus/visual/eyes/model/AccessibilityCheckResult.java:32]: discussed

File: vividus-plugin-applitools/src/main/resources/applitools-visual-comparison.ftl (hunk 11 above)
Comment (uarlouski) [reply vividus-plugin-applitools/src/main/resources/applitools-visual-comparison.ftl:19]: discussed

**Review by valfirst — APPROVED** [review]

**Pull request merged**